dali/driver/hasseb.py

```diff
@@ -1,128 +1,193 @@
-        
-        needsResponse = command.response is not None
-        a, b = command.frame.as_byte_sequence
-        
-        # print(u" SEND: a=0x%x b=0x%x"%(a,b))
-        
-        backward = self._writeDali( a, b, needsResponse)
-        if command._response:
-            response = command._response(backward)
+                return BackwardFrameError(255)
+        self.logger.error("Invalid Frame")
+        return None
+
+class SyncHassebDALIUSBDriver(HassebDALIUSBDriver, SyncDALIDriver):
+    """Synchronous ``DALIDriver`` implementation for Hasseb DALI USB device.
+    """
+
+    def __init__(self, bus=None, port_numbers=None, interface=0):
+        self.backend = USBBackend(
+            HASSEB_USB_VENDOR,
+            HASSEB_USB_PRODUCT,
+            bus=bus,
+            port_numbers=port_numbers,
+            interface=interface
+        )
+
+    def send(self, command, timeout=2000):
+        self.backend.write(self.construct(command))
+        frame = None
+        backoff = 0.010
+
+        if command.response is not None:
+            for i in range(7):
+                frame = self.extract(self.backend.read(timeout=timeout))
+                if isinstance(frame, HassebDALIUSBNoAnswer):
+                    self.backend.write(self.construct(command))
+                if isinstance(frame, BackwardFrame):
+                    if command.response:
+                        return command.response(frame)
+                    return frame
+                backoff += backoff
+                sleep(backoff)
+            raise MissingResponse()
+        return None
+
+class AsyncHassebDALIUSBDriver(HassebDALIUSBDriver, AsyncDALIDriver):
+    """Asynchronous ``DALIDriver`` implementation for Hasseb DALI USB device.
+       This driver is FAKE, since Hasseb does NOT support async communication.
+    """
+    _pending = None
+
+    def __init__(self, bus=None, port_numbers=None, interface=0):
+        self.backend = USBListener(
+            self,
+            HASSEB_USB_VENDOR,
+            HASSEB_USB_PRODUCT,
+            bus=bus,
+            port_numbers=port_numbers,
+            interface=interface
+        )
+
+    def send(self, command, callback=None, **kw):
+        data = self.construct(command)
+        if command.response is not None:
+            self._pending = command, callback, kw
+        else:
+            self._pending = None
+        self.backend.write(data)
+
+    def receive(self, data):
+        frame = self.extract(data)
+        if isinstance(frame, HassebDALIUSBNoDataAvailable):
+            return
+        elif isinstance(frame, BackwardFrame):
+            if self._pending:
+                command, callback, kw = self._pending
+                callback(command.response(frame), **kw)
+            else:
+                logger.error("Received frame for no pending command")
         else:
-            response = None
+            logger.error("Received frame is not BackwardFrame")
```

Comment:
Can you please add a comment that here's the place to handle incoming forward frames?

Reply:
Never dealt with a DALI install beyond mine. 

Is it possible to have more than one controller ? (Does it make sense for a controller to receive a ForwardFrame) ?

Reply:
Sure it's possible to have more than one controller. DALI is multi master. Everything which comes from even a simple DALI switch sends forward frames.

Reply:
It can't receive unsolicited frames no... multi master with Hasseb would become a big no no...